cmd/root.go

```diff
@@ -48,20 +48,19 @@ var rootCmd = &cobra.Command{
 		env.InitLogger()
 
 		source := env.Config().GetString("Source")
-		outputDir := env.Config().GetString("OutputDir")
 
 		configs := []ocp.Translator{}
-		ocpMaster := new(ocp.ConfigMaster)
-		ocpNode := new(ocp.ConfigNode)
-		ocpMaster.Add(source)
-		ocpNode.Add(source)
-		configs = append(configs, ocpMaster, ocpNode)
+		ocpOAuth := new(ocp.OAuthConfig)
+		ocpSDN := new(ocp.SDNConfig)
+		ocpOAuth.Add(source)
+		ocpSDN.Add(source)
+
+		configs = append(configs, ocpOAuth, ocpSDN)
```

Comment:
root.go should not be creating these configs. I think you should have something like a `cpma.go` with a `Start` method that is launched from `root.go`. The `cpma.go` should be passed configurations based on the CLI parameters and start the process of doing migrations.

The `root.go` file should be relegated to handling CLI parameters then passing those to the `cpma.go`. Nothing more. Right now `root.go` has too much knowledge of the process.

Reply:
I agree with that meanwhile others in the group don't.
So if you're against it please speak now or never!

Reply:
Yeah `root.go` should be parameters from CLI and then pass to a single entry that does all the work. I just like to separate the concerns as much as possible. Like I could theoretically rip out root.go and still be able to call the thing that does all the work from some other project, it's probably not a goal but it doesn't cost much to have a different file called from root.go. IMO

Reply:
That's exactly what I thought.
The reason it that apparently it's the practice when using "github.com/spf13/cobra".

Reply:
> That's exactly what I thought.
> The reason it that apparently it's the practice when using "github.com/spf13/cobra".

I disagree, cobra does not dictate how you call the rest of your app. Let's be clear here, I'm not debating the need for the `root.go` file. And if cobra requires it to be called `root.go` that is also okay. What I'm against is this code living in `root.go`.

```golang
		ocpOAuth := new(ocp.OAuthConfig)
		ocpSDN := new(ocp.SDNConfig)
		ocpOAuth.Add(source)
		ocpSDN.Add(source)

		configs = append(configs, ocpOAuth, ocpSDN)

		for _, config := range configs {
			config.Extract()
			config.Transform()
			ocp.DumpManifests(config.GenYAML())
}
```

I'm of the opinion this should be in the core of your application. IMO `root.go` is NOT the core of your application. It is merely the "UI" for this application. Basically the struct that knows how to handle CLI parameters really doesn't care about what the core of the application needs. 

Basically I know to look in `cmd/` to see what parameters there are and how they are parsed. And I look to see how the application is started. 

So to satisfy the request I'm making it is as simple as moving the above code to another file in the `pkg/` directory, something like this:

```golang
MigrationAssistant struct {
 // the fields required to handle the transformations etc, I would defer to the POC that @eriknelson has been working on for this
}

func (m *MigrationAssistant) Start(params map[string]string) {
  // take in the parameters you need to start this application
}
```

Then in `root.go` I would do the following:

```golang
var rootCmd = &cobra.Command{
	Use:   "cpma",
	Short: "Helps migration cluster configuration of a OCP 3.x cluster to OCP 4.x",
	Long:  `Helps migration cluster configuration of a OCP 3.x cluster to OCP 4.x`,
	Run: func(cmd *cobra.Command, args []string) {
		env.InitConfig()
		env.InitLogger()

		// start the application
		ma := MigrationAssistant{}
		ma.Start(PASS_YOUR_PARAMS_HERE)
},
...
```

Do you see now? Basically `root.go` knows all about parsing parameters. It knows how to transform those CLI parameters into something that can be passed into the `Start` of the application. It's cleaner and easier to see where the stuff for cobra is and the stuff required by your application.

Reply:
I think that's going to clarify it for good. Thanks